### PR TITLE
fix for low decimals tokens

### PR DESCRIPTION
**Pull request by LouisMeMyself**

### CODE REVIEW

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 1]
```diff
@@ -108,6 +113,13 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
         MCJ = _MCJ;
         isNative = _isNative;
         poolInfo = PoolInfo({lastRewardTimestamp: block.timestamp, accTokenPerShare: 0});
+
+        JoePair pair = JoePair(address(_lpToken));
+        IERC20Metadata token0 = IERC20Metadata(pair.token0());
+        IERC20Metadata token1 = IERC20Metadata(pair.token1());
+
+        if (_tokenPerSec > 10**6) ACC_TOKEN_PRECISION = 10**(36 - (token0.decimals() + token1.decimals()) / 2);
```
Comment (jummy123) [inline contracts/rewarders/SimpleRewarderPerSec.sol:121]: Don't forget to do this check/calculation again in `setRewardRate`

Comment (LouisMeMyself) [reply contracts/rewarders/SimpleRewarderPerSec.sol:121]: Really good catch, this actually involves some changes on `accTokenPerShare` too !

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 2]
```diff
@@ -81,7 +94,12 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
     mapping(address => UserInfo) public userInfo;
 
     uint256 public tokenPerSec;
-    uint256 private constant ACC_TOKEN_PRECISION = 1e12;
+
+    uint256 private accTokenPrecision;
+    // @dev decimals of pair
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:99]: `decimals`->`Decimals`

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 3]
```diff
@@ -81,7 +94,12 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
...
+    uint256 private pairDecimals;
+    // @dev decimals of reward token
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:101]: `decimals`->`Decimals`

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 4]
```diff
@@ -108,6 +126,55 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
...
+        //        = 1e46
+        //
+        // Edge case n4:
+        // `lpSupply` is in 18 decimals and is very big, 1e(12+18)
+        // `tokenPerSec` is very small, i.e 1 WEI
+        // `updatePool` is updated every 1s, i.e. 1 seconds
+        // decimals of result:
+        // result = 1 * 1 * 1e(19 + 18 - 1) / 1e30
+        //        = 1 * 1 * 1e36 / 1e30
+        //        = 1e6
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:168]: These need to be explained in simpler terms

**Review by cryptofish7 — CHANGES_REQUESTED** [review]
We need unit tests too

**Comment by jummy123** [comment]
So fix is just to change `ACC_TOKEN_PRECISION` value?

**Comment by jummy123** [comment]
Would be nice to have a unit test showing the behavior that was failing before.

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 5]
```diff
@@ -48,6 +48,14 @@ interface IMasterChefJoe {
...
  * 100,000 XYZ and set the block reward accordingly so it's fully distributed after 30 days.
  *
+ *
+ * Issue with previous version is that this can return 0 or be very inacurate with some tokens:
+ *      uint256 timeElapsed = block.timestamp.sub(pool.lastRewardTimestamp);
+ *      uint256 tokenReward = timeElapsed.mul(tokenPerSec);
+ *      accTokenPerShare = accTokenPerShare.add(
+ *          tokenReward.mul(ACC_TOKEN_PRECISION).div(lpSupply)
+ *      );
+ *  The goal with those changes is to prevent this, without any overflow too.
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:58]:
```
 * Issue with the previous version is that this fraction, `tokenReward.mul(ACC_TOKEN_PRECISION).div(lpSupply)`, 
 * can return 0 or be very inacurate with some tokens:
 *      uint256 timeElapsed = block.timestamp.sub(pool.lastRewardTimestamp);
 *      uint256 tokenReward = timeElapsed.mul(tokenPerSec);
 *      accTokenPerShare = accTokenPerShare.add(
 *          tokenReward.mul(ACC_TOKEN_PRECISION).div(lpSupply)
 *      );
 *  The goal is to set ACC_TOKEN_PRECISION high enough to prevent this without causing overflow too.
 ```

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 6]
```diff
@@ -108,6 +117,22 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
         MCJ = _MCJ;
         isNative = _isNative;
         poolInfo = PoolInfo({lastRewardTimestamp: block.timestamp, accTokenPerShare: 0});
+
+        // Edge case n1: maximize the numerator, minimize the denominator.
+        // `lpSupply` = 1 WEI
+        // `tokenPerSec` = 1e(30)
+        // `timeElapsed` = 31 years, i.e. 1e9 seconds
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:124]:
```
// Given the fraction, tokenReward * ACC_TOKEN_PRECISION / lpSupply, we consider
// several edge cases.

// Edge case n1: maximize the numerator, minimize the denominator.
...
```

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 7]
```diff
@@ -108,6 +117,22 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
...
+        // result = 1e9 * 1e30 * 1e36 / 1
+        //        = 1e75
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:126]:
```
//        = 1e75
// (No overflow as max uint256 is 1.15e77).
...
```

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 8]
```diff
@@ -108,6 +117,22 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
...
+        //
+        // Edge case n2: minimize the numerator, maximize the denominator.
+        // `lpSupply` = 1e30
+        // `tokenPerSec` = 1 WEI
+        // `timeElapsed` = 1 second
+        // decimals of result:
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:132]: Remove this line, `// decimals of result:`

File: contracts/rewarders/SimpleRewarderPerSec.sol [hunk 9]
```diff
@@ -108,6 +117,22 @@ contract SimpleRewarderPerSec is IRewarder, BoringOwnable, ReentrancyGuard {
...
+        // result = 1 * 1 * 1e36 / 1e30
+        //        = 1e36
```
Comment (cryptofish7) [inline contracts/rewarders/SimpleRewarderPerSec.sol:134]:
```
//        = 1e36
// (Not rounded to zero, therefore ACC_TOKEN_PRECISION = 1e36 is safe).

**Review by cryptofish7 — CHANGES_REQUESTED** [review]

**Review by cryptofish7 — APPROVED** [review]
